### PR TITLE
Optimize service retrieval for inlineability.

### PR DESCRIPTION
**Pull request by cramforce**
Functions such as `viewportFor(win)` can now be inlined as `win.services.viewport.obj`.

Changes the method `getWin()` to use a property access in all places

- Main trade off is that caching the window will break when a DOM node moves windows. That should never happen, though :)
- Closure Compiler knows that property access is side effect free (we promised it), so it can more aggressively inline.
- This should actually be a major speed up in itself, as it avoid 2 trips into C++ per window lookup.

Great idea @erwinmombay, @dvoytenko

### CODE REVIEW

File: src/service.js [hunk 1]
```diff
@@ -32,6 +33,38 @@ import {dev} from './log';
...
 /**
+ * Returns a service with the given id. Assumes that it has been constructed
+ * already.
+ * @param {!Window} win
+ * @return {!Object} The service.
+ */
+export function getExistingServiceForWindow(win, id) {
+  if (getMode().localDev) {
+    dev.assert(
+       win.services && win.services[id] && win.services[id].obj,
```
Comment (jridgewell) [inline src/service.js]: Is it necessary to guard? The TypeError would serve the purpose as well, just without the message.

**Comment by erwinmombay** [comment]
i can't take any credit, was all you and @dvoytenko :P

**Comment by erwinmombay** [comment]
Fixes https://github.com/ampproject/amphtml/issues/4096

File: src/service.js [hunk 2]
```diff
@@ -32,6 +33,38 @@ import {dev} from './log';
 let ServiceHolderDef;
 
 /**
+ * Returns a service with the given id. Assumes that it has been constructed
+ * already.
+ * @param {!Window} win
+ * @return {!Object} The service.
```
Comment (erwinmombay) [inline src/service.js]: missing id param

File: src/service.js [hunk 3]
```diff
@@ -32,6 +33,38 @@ import {dev} from './log';
...
+       win.services && win.services[id] && win.services[id].obj,
+        'Window service does not exist %s', id);
+  }
+  return win.services[id].obj;
+}
+
+/**
+ * Returns a service with the given id. Assumes that it has been constructed
+ * already.
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
```
Comment (erwinmombay) [inline src/service.js]: missing id param

File: src/service.js (hunk 1 above)
Comment (cramforce) [inline src/service.js]: Well, with that one could kill the entire assert but it produces a much nicer error. Should all be optimized away.

**Comment by cramforce** [comment]
Unfortunately some call sites are not being inlined. Investigating https://github.com/google/closure-compiler/issues/1902

File: src/service.js (hunk 1 above)
Comment (dvoytenko) [inline src/service.js]: Why check for `localDev`? It'd be good if it failed the same in tests as well. And we remove `dev.assert` in PROD.

**Comment by dvoytenko** [comment]
@cramforce All looks good. Pending the inlineability.

**Comment by cramforce** [comment]
This PR is now much, MUCH bigger :(

I switched `.getWin()` to `.win` everywhere. This is also likely significantly faster and it allows much more stuff to be inlined, because the compiler knows that the expression is side effect free.

**Comment by cramforce** [comment]
As noted in the description: I've been meaning to do this anyway, because "This should actually be a major speed up in itself, as it avoid 2 trips into C++ per window lookup."

File: extensions/amp-accordion/0.1/amp-accordion.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ class AmpAccordion extends AMP.BaseElement {
     this.sections_ = this.getRealChildren();
 
     /** @const @private {!Window} */
-    this.win_ = this.getWin();
+    this.win_ = this.win;
```
Comment (dvoytenko) [inline extensions/amp-accordion/0.1/amp-accordion.js]: Grrr. `this.win_` should probably be removed completely. Ok to pass on this for this PR though.

File: extensions/amp-live-list/0.1/amp-live-list.js [hunk 5]
```diff
@@ -113,7 +113,7 @@ export class AmpLiveList extends AMP.BaseElement {
   /** @override */
   buildCallback() {
     /** @const {!Window} */
-    this.win = this.getWin();
+    this.win = this.win;
```
Comment (dvoytenko) [inline extensions/amp-live-list/0.1/amp-live-list.js]: Remove completely?

File: src/service/ampdoc-impl.js [hunk 6]
```diff
@@ -158,18 +159,18 @@ export class AmpDocSingle extends AmpDoc {
...
     super();
-    /** @private @const {!Window} */
+    /** @public @const {!Window} */
     this.win = win;
   }
 
   /** @override */
-  isSingleDoc() {
-    return true;
+  getWin() {
```
Comment (dvoytenko) [inline src/service/ampdoc-impl.js]: Do we need this method now?

**Comment by dvoytenko** [comment]
@cramforce 

> I switched .getWin() to .win everywhere.

This is certainly ok. But how meaningful is this? (Aside from removing calls to `defaultView`) aren't the `getWin()` fully inlineable anyway?

File: src/base-element.js [hunk 7]
```diff
@@ -163,14 +166,17 @@ export class BaseElement {
     return this.layout_;
   }
 
-  /** @protected @return {!Window} */
+  /**
+   * DO NOT CALL. Retained for backward compat during rollout.
+   * @protected @return {!Window}
```
Comment (jridgewell) [inline src/base-element.js]: `@deprecated`?

File: extensions/amp-live-list/0.1/amp-live-list.js (hunk 5 above)
Comment (cramforce) [inline extensions/amp-live-list/0.1/amp-live-list.js]: Done

File: src/base-element.js (hunk 7 above)
Comment (cramforce) [inline src/base-element.js]: Its protected by presubmit. It isn't really deprecated :)

File: src/service.js (hunk 3 above)
Comment (cramforce) [inline src/service.js]: Done

File: extensions/amp-live-list/0.1/amp-live-list.js (hunk 5 above)
Comment (cramforce) [inline extensions/amp-live-list/0.1/amp-live-list.js]: Done

**Comment by cramforce** [comment]
We currently can never inline automatically `this.getWin()` due to lack of typing and generally not allowing it for non private properties in our current model..

**Comment by erwinmombay** [comment]
@cramforce do we need to add win to extern?

**Comment by erwinmombay** [comment]
ah nvm, forgot public props should be safe across compilation units

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 8]
```diff
@@ -37,7 +37,7 @@ export class AmpSlideScroll extends BaseCarousel {
   /** @override */
   buildCarousel() {
     /** @private @const {!Window} */
-    this.win_ = this.getWin();
+    this.win_ = this.win;
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slidescroll.js]: Another case with `this.win_`.

File: extensions/amp-sidebar/0.1/amp-sidebar.js [hunk 9]
```diff
@@ -37,7 +37,7 @@ export class AmpSidebar extends AMP.BaseElement {
   /** @override */
   buildCallback() {
     /** @private @const {!Window} */
-    this.win_ = this.getWin();
+    this.win_ = this.win;
```
Comment (dvoytenko) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: Another `this.win_`

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 10]
```diff
@@ -90,7 +90,7 @@ export class AmpUserNotification extends AMP.BaseElement {
   createdCallback() {
 
     /** @private @const {!Window} */
-    this.win_ = this.getWin();
+    this.win_ = this.win;
```
Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: And one more. Totally fine to defer, but let's make sure this doesn't get lost.

Comment (erwinmombay) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: ill file an issue and clean it up

Comment (erwinmombay) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: files here https://github.com/ampproject/amphtml/issues/4138

**Comment by cramforce** [comment]
Is this LG?

**Comment by cramforce** [comment]
Friendly ping :)

**Comment by erwinmombay** [comment]
@cramforce LGTM!

**Comment by zhouyx** [comment]
@cramforce Could you please explain on `2 trips into C++ per window lookup`?

**Comment by cramforce** [comment]
Yes, happy to!
The browser is written in C++, we write JavaScript. It is relatively (don't
get crazy about it, but this here we do a lot) to call from Js into C++ and
vice versa. Its more than just a function call, things have to be setup for
the other side expectation. This makes it relatively slow and avoiding such
jump can pay off when they are done a lot.

**Pull request merged**